### PR TITLE
Improve RetrieveChat

### PR DESCRIPTION
**Pull request by thinkall**
## Why are these changes needed?

- Use upsert in batch of 40000 to improve chromadb upsert stability, avoid error for super large corpus.
- Improve `Update Context` logic, support customizing trigger words `customized_answer_prefix`. If `customized_answer_prefix` is given and it is not in the answer, `Update Context` will also be triggered.
- Add a parameter `no_update_context` to stop `Update Context`, useful for ablation experiments.
- Add an example of customizing prompt, trigger word and few-shot learning to `RetrieveChat` notebook.
- Update tests.

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by sonichi** [comment]
This PR decreases coverage. Can we avoid that?

**Comment by thinkall** [comment]
> This PR decreases coverage. Can we avoid that?

I see `test/agentchat/test_retrievechat.py` is skipped, is there any concerns for these tests?

**Comment by sonichi** [comment]
> > This PR decreases coverage. Can we avoid that?
> 
> I see `test/agentchat/test_retrievechat.py` is skipped, is there any concerns for these tests?

Some tests are skipped because chromadb is not installed in [test]. We can add it.
The test that depends on OpenAI should check whether openai is installed and skip if not, like other tests.

File: autogen/agentchat/contrib/retrieve_user_proxy_agent.py
Comment (sonichi) [inline autogen/agentchat/contrib/retrieve_user_proxy_agent.py:1]: There are changes beyond the upsert in batch. Could you please add more explanation in the PR and try to document them more clearly in the code? If the added functionality can be illustrated by a notebook example it'll be the best.

Comment (thinkall) [reply autogen/agentchat/contrib/retrieve_user_proxy_agent.py:1]: Added

File: notebook/agentchat_RetrieveChat.ipynb
Comment (sonichi) [inline notebook/agentchat_RetrieveChat.ipynb:1]: For the new example, could you explain what happened here, when no "UPDATE CONTEXT" is sent but the context is still updated? 

assistant (to ragproxyagent):

I'm sorry, I do not have enough information about North Marion High School and Seoul High School to provide an answer. Please provide more context or information about the schools.

--------------------------------------------------------------------------------
Updating context and resetting conversation.

Comment (thinkall) [reply notebook/agentchat_RetrieveChat.ipynb:1]: This is triggered by the customized trigger words `customized_answer_prefix`. If `customized_answer_prefix` is given and it is not in the answer, `Update Context` will also be triggered. It's `the answer is` in the notebook.

**Review by rickyloynd-microsoft — APPROVED** [review]

**Comment by sonichi** [comment]
One last change we need: bump the version of autogen and the version number in the notebook. Then I'll need to make a new release. Otherwise the notebook will not work.

**Comment by thinkall** [comment]
> One last change we need: bump the version of autogen and the version number in the notebook. Then I'll need to make a new release. Otherwise the notebook will not work.

Updated to 0.1.2

**Review by pcdeadeasy — CHANGES_REQUESTED** [review]
I saw a chunk of code that was commented out in generate_init_message.
Also consider making _generate_retrieve_user_reply more readable.

File: autogen/agentchat/contrib/retrieve_user_proxy_agent.py [hunk 1]
```diff
@@ -230,6 +230,11 @@ def _generate_retrieve_user_reply(
         sender: Optional[Agent] = None,
         config: Optional[Any] = None,
     ) -> Tuple[bool, Union[str, Dict, None]]:
+        """In this function, we will update the context and reset the conversation based on different conditions.
+        We'll update the context and reset the conversation if no_update_context is False and either of the following:
```
Comment (sonichi) [inline autogen/agentchat/contrib/retrieve_user_proxy_agent.py:234]: It'll be easier to think of "update_context is True" than "no_update_context is False".

**Review by pcdeadeasy — APPROVED** [review]
Changes look good to me. once suggestion that you could consider is to rename the rerieve_user_proxy_agent.py as **RAG**_user_proxy_agent.py.

**Pull request merged**